Add `AlphaAndBlend` operation to all chapters

pavpanchekha opened the pull request:
In Chapter 11, we first add `Alpha` and `Blend` separately, and then merge them later to create `AlphaAndBlend`. The `AlphaAndBlend` operation gains an extra `isolate` parameter so that a layer is created when clipping with destination-in. Other chapters are fixed up to use `AlphaAndBlend` instead of `SaveLayer`.

chrishtr commented on book/visual-effects.md:
````diff
@@ -1079,31 +1079,32 @@ which describes the opacity as a floating-point number from 0 to 1.
-    def __init__(self, sk_paint, children):
-        self.sk_paint = sk_paint
+``` {.python}
+class Alpha:
````
Sorry for the late idea but should we call it Opacity instead of Alpha? Alpha is not the same as opacity..

pavpanchekha replied:
Sure, but then OpacityAndBlend sounds bad, can I just call the final thing Blend?

chrishtr replied:
Sounds good!

chrishtr commented on book/visual-effects.md:
```diff
@@ -1202,8 +1203,8 @@ class Pixel:
 
-That `Alphaf` operation applies to pixels in one surface. But with
-`SaveLayer` we will end up with two surfaces, with all of their pixels
+That `alphaf` operation applies to pixels in one surface. But with
```
```suggestion
That `Alphaf` operation applies to pixels in one surface. But with
```

pavpanchekha replied:
Ah, I see, you meant the Skia parameter name. Got it.

chrishtr commented on book/visual-effects.md:
```diff
@@ -1558,56 +1591,83 @@ the page, so that clipping only applies to that element.
-            SaveLayer(skia.Paint(Alphaf=opacity), cmds,
-                should_save=needs_opacity),
-            SaveLayer(skia.Paint(BlendMode=skia.kDstIn), [
+        Blend(blend_mode, clip_radius > 0, [
```
We could avoid the isolation parameter by this; WDYT?

if not blend_mode and clip_radius > 0:
  blend_mode = "source-over"
Blend(blend-mode, ...)

pavpanchekha replied:
Oh yeah, we can! I initially had the parse elsewhere and changed it for unrelated reasons but now this works great!

chrishtr commented on src/lab11.py:
```diff
@@ -70,30 +70,73 @@ def parse_blend_mode(blend_mode_str):
-        self.should_save = should_save
-        self.should_paint_cmds = should_paint_cmds
-        self.sk_paint = sk_paint
+class Alpha:
```
I guess these classes are only there to make it easier to refer to them in the chapter. But it's dead code, which is something we don't have anywhere else...

pavpanchekha replied:
Yeah, I can put them in the example file for compare to do it’s thing.

chrishtr commented on src/lab13.py:
```diff
@@ -760,15 +777,13 @@ def paint_visual_effects(node, cmds, rect):
-
-    node.save_layer = save_layer
+    blend_op = AlphaAndBlend(opacity, blend_mode, False, node, [
+        ClipRRect(rect, clip_radius,
```
This needs to be AlphaOrBlend("destination-in", ...

chrishtr replied:
Can we delete ClipRRect also?

pavpanchekha replied:
Will do but I think separate PR, it is a little complicated due to the local absolute thing in a later chapter. (I think we want a ClipRRect operation even if it works by destination in blending.

chrishtr replied:
sg

pavpanchekha commented:
This blocks a PR about removing ClipRRect, which is going to be moderately nasty, so I'd appreciate quick review. I think I've addressed all of the issues noted previously.
